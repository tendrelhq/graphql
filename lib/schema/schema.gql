enum AuthenticationProvider {
  legacy
  clerk
}

enum LanguageCode {
  en
  es
}

type Language {
  id: ID!
  code: LanguageCode!
  name: Name!
  name_id: ID!
}

type Name {
  id: ID!
  language: Language!
  language_id: ID!
  value: String!
}

type Tag {
  id: ID!
  name: Name!
  name_id: ID!
  type: String!
  parent: Tag
  parent_id: ID
}

type Customer {
  id: ID!
  active: Boolean!
  name: Name!
  name_id: ID!
  defaultLanguage: Language!
  default_language_id: ID!
}

type Location {
  id: ID!
  active: Boolean!
  activated_at: String
  deactivated_at: String
  name: Name!
  name_id: ID!
  parent: Location
  parent_id: ID
  scan_code: ID
  site: Location!
  site_id: ID!
  tags: [Tag!]!

  children(options: LocationsQueryOptions): [Location!]!
}

type User {
  id: ID!
  active: Boolean!
  activated_at: String
  deactivated_at: String
  authentication_provider: Tag
  authentication_provider_id: ID
  language: Language!
  language_id: ID!
  name: String!
  tags: [Tag!]!
}

type Worker {
  id: ID!
  active: Boolean!
  activated_at: String
  deactivated_at: String
  language: Language!
  language_id: ID!
  role: Tag!
  role_id: ID!
  scan_code: String
  tags: [Tag!]!
  user: User!
  user_id: ID!
}

type Query {
  customers: [Customer!]!
  customer(id: ID!): Customer!
  languages: [Language!]!
  locations(customerId: ID!, options: LocationsQueryOptions): [Location!]!
  location(id: ID!): Location!
  roles: [Tag!]!
  workers(customerId: ID!, options: WorkersQueryOptions): [Worker!]!
}

input LocationsQueryOptions {
  cornerstone: Boolean
  site: Boolean
}

input WorkersQueryOptions {
  active: Boolean
  site: ID
}

type Mutation {
  # Location CRUD
  createLocation(input: CreateLocationInput!): Location!
  updateLocation(input: UpdateLocationInput!): Location!

  # User
  ## CRUD
  createUser(input: CreateUserInput!): User!
  updateUser(input: UpdateUserInput!): User!
  ## Helpers
  acceptInvite(input: AcceptInviteInput!): User!

  # Worker CRUD
  createWorker(input: CreateWorkerInput!): Worker!
  updateWorker(input: UpdateWorkerInput!): Worker!
}

input CreateLocationInput {
  id: ID
  active: Boolean!
  name: CreateNameInput!
  org_id: ID!
  scan_code: ID
}

input UpdateLocationInput {
  id: ID!
  activated_at: String
  deactivated_at: String
  name: UpdateNameInput
  scan_code: ID
}

input CreateUserInput {
  id: ID
  active: Boolean!
  authentication_provider_id: ID
  authentication_identity: ID
  language_id: ID!
  name: String!
}

input UpdateUserInput {
  id: ID!
  authentication_provider_id: ID
  authentication_identity: ID
  language_id: ID!
  name: String!
}

input CreateWorkerInput {
  id: ID
  active: Boolean!
  language_id: ID!
  org_id: ID!
  role_id: ID!
  scan_code: ID
  user_id: ID!
}

input UpdateWorkerInput {
  id: ID!
  activated_at: String
  deactivated_at: String
  language_id: ID!
  role_id: ID!
  scan_code: ID
}

input CreateNameInput {
  id: ID
  active: Boolean!
  language_id: ID!
  value: String!
}

input UpdateNameInput {
  id: ID!
  activated_at: String
  deactivated_at: String
  language_id: ID!
  value: String!
}

input AcceptInviteInput {
  id: ID!
  authentication_identity: ID!
}