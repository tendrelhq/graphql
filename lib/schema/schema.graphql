enum Language {
  en
  es
}

type Tag {
  id: ID!
  name: String!
  type: String!
  parent: ID
}

type Customer {
  id: ID!
  name: String!
  defaultLanguage: Language!
}

type Location {
  id: ID!
  customerId: ID!
  name: String!
  parentId: ID
  tags: [Tag!]!
}

type Query {
  customers: [Customer!]!
  customer(id: ID!): Customer!
  locations(customerId: ID!, parentId: ID): [Location!]!
  sites(customerId: ID!): [Location!]!
}

type Mutation {
  createCustomer(input: CreateCustomerInput!): ID!
  updateCustomer(input: UpdateCustomerInput!): ID!
  deleteCustomer(id: ID!): ID!
  createLocation(input: CreateLocationInput!): ID!
  updateLocation(input: UpdateLocationInput!): ID!
  deleteLocation(id: ID!): ID!
}

input CreateTagInput {
  id: ID!
  name: String!
  type: String!
  parent: ID
}

input UpdateTagInput {
  id: ID!
  name: String
}

input CreateCustomerInput {
  id: ID!
  name: String!
  defaultLanguage: Language!
}

input UpdateCustomerInput {
  id: ID!
  name: String
  defaultLanguage: Language
}

input CreateLocationInput {
  id: ID!
  customerId: ID!
  name: String!
  tags: [CreateTagInput!]!
}

input UpdateLocationInput {
  id: ID!
  name: String
  tags: [UpdateTagInput!]
}
