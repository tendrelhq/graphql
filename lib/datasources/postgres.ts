import type { Request } from "express";
import z from "myzod";
import postgres, { type Fragment } from "postgres";
import { makeActiveLoader } from "./activatable";
import makeAttachmentLoader from "./attachment";
import { makeAuditableLoader } from "./auditable";
import makeCustomerRequestedLanguageLoader from "./crl";
import { makeDescriptionLoader } from "./description";
import { makeDynamicStringLoader } from "./dynamicString";
import makeInvitationLoader from "./invitation";
import makeLanguageLoader from "./language";
import makeLocationLoader from "./location";
import {
  makeDisplayNameLoader,
  makeNameLoader,
  makeNameMetadataLoader,
} from "./name";
import makeOrganizationLoader from "./organization";
import { makeRequirementLoader } from "./requirement";
import { makeSopLoader } from "./sop";
import { makeStatusLoader } from "./status";
import makeTagLoader from "./tag";
import makeUserLoader from "./user";
import makeWorkerLoader from "./worker";

const { DB_MAX_CONNECTIONS, DB_STATEMENT_TIMEOUT } = z
  .object({
    DB_MAX_CONNECTIONS: z.number({ coerce: true }).default(3),
    DB_STATEMENT_TIMEOUT: z.number({ coerce: true }).default(10),
  })
  .parse(process.env, { allowUnknown: true });

export const sql = postgres({
  max: DB_MAX_CONNECTIONS,
  // TODO: this is probably a good idea...
  // connection: {
  //   statement_timeout: DB_STATEMENT_TIMEOUT * 1000, // milliseconds
  // },
});

export type SQL = typeof sql;
export type TxSQL = Parameters<Parameters<typeof sql.begin>[1]>[0];

/**
 * Like Array.prototype.join but for sql.Fragments.
 */
export function join(xs: readonly Fragment[], d: Fragment) {
  return xs.reduce((acc, x, i) => sql`${acc} ${i ? sql`${d} ${x}` : x}`, sql``);
}

/**
 * Type predicate for whether to include a user input in a dynamic update
 * clause. Note that we only check for `undefined`, and not `null`, because we
 * say that `null` indicates we want to set the underlying database column to
 * NULL.
 */
export function shouldUpdate<T>(input?: T, existing?: T): input is T {
  return typeof input !== "undefined" && input !== existing;
}

export function unionAll(xs: readonly Fragment[]) {
  return join(xs, sql`UNION ALL`);
}

export function orm(req: Request) {
  return {
    active: makeActiveLoader(req),
    attachment: makeAttachmentLoader(req),
    auditable: makeAuditableLoader(req),
    crl: makeCustomerRequestedLanguageLoader(req),
    description: makeDescriptionLoader(req),
    displayName: makeDisplayNameLoader(req),
    dynamicString: makeDynamicStringLoader(req),
    invitation: makeInvitationLoader(req),
    language: makeLanguageLoader(req),
    location: makeLocationLoader(req),
    name: makeNameLoader(req),
    nameMetadata: makeNameMetadataLoader(req),
    organization: makeOrganizationLoader(req),
    requirement: makeRequirementLoader(req),
    sop: makeSopLoader(req),
    status: makeStatusLoader(req),
    tag: makeTagLoader(req),
    user: makeUserLoader(req),
    worker: makeWorkerLoader(req),
  };
}

export type ORM = ReturnType<typeof orm>;
