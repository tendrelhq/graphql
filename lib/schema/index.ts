export * from "./__generated__/resolvers.generated";
export * from "./__generated__/typeDefs.generated";

export type Context = {
  authScope?: string;
  languageTypeId: number;
};
