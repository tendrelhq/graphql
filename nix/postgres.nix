{inputs, ...}: {
  imports = [
    inputs.process-compose.flakeModule
  ];
  perSystem = {
    config,
    self',
    pkgs,
    lib,
    ...
  }: {
    devShells.postgres = pkgs.mkShellNoCC {
      name = "tendrelhq/graphql+postgres";
      buildInputs = [
        config.packages.postgresql
        pkgs.perlPackages.TAPParserSourceHandlerpgTAP # pg_prove
        pkgs.python3.pkgs.sqlfmt
        pkgs.sqitchPg
        pkgs.squawk
      ];
    };

    packages = {
      inherit (pkgs) postgresql_14; # beta is on 14.12
      postgresql = pkgs.postgresql_16;
      pgddl = pkgs.stdenv.mkDerivation rec {
        name = "pgddl";
        version = "0.29";
        src = pkgs.fetchFromGitHub {
          owner = "lacanoid";
          repo = name;
          rev = version;
          hash = "sha256-W3G6TGtkj+zXXdGZZR0bmZhsLuFJvuGTlDoo8kL8sf0=";
        };
        buildInputs = with pkgs; [perl config.packages.postgresql];
        postPatch = ''
          patchShebangs .
        '';
        installPhase = ''
          install -D -t $out/share/postgresql/extension *.sql
          install -D -t $out/share/postgresql/extension *.control
        '';
      };
    };

    process-compose.devenv = {
      imports = [
        inputs.services.processComposeModules.default
      ];

      cli.options.no-server = false;

      services.postgres.pg1 = {
        enable = true;
        package = config.packages.postgresql;
        #
        extensions = exts:
          with exts;
          with config.packages; [
            pg_cron
            pgddl
            pgtap
            plpgsql_check
          ];
        hbaConf = lib.mkForce [
          # This one is the default docker bridge network.
          {
            type = "host";
            database = "all";
            user = "all";
            address = "172.17.0.0/16";
            method = "trust";
          }
          # On my machine this is the tendrel_default network. YMMV.
          {
            type = "host";
            database = "all";
            user = "all";
            address = "172.18.0.0/16";
            method = "trust";
          }
        ];
        initialDatabases = [
          {
            name = "dev";
          }
        ];
        settings = {
          listen_addresses = "0.0.0.0"; # so docker0 can connect
          log_statement = "all";
          logging_collector = false;
          shared_preload_libraries = "pg_cron,plpgsql_check";
        };
      };

      services.pgadmin.pga1 = {
        enable = true;
        extraConfig = {
          # Run in desktop mode with the master password disabled.
          MASTER_PASSWORD_REQUIRED = false;
          SERVER_MODE = false;
        };
        initialEmail = "";
        initialPassword = "";
      };

      settings.processes = {
        pga1 = {
          depends_on.pg1.condition = "process_healthy";
        };
      };
    };
  };
}
