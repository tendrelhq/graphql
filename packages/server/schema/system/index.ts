import z from "zod";

const GlobalId = z.string().transform(id => {
  const parts = id.split(":");
  if (parts.length < 2) {
    throw "invariant violated: invalid global identifier";
  }
  return {
    type: parts[0],
    id: parts[1],
    suffix: parts.slice(2),
  };
});

/**
 * Global Identifier. Externally it just an opaque string. Internally it may
 * have some meaning. Right now this is just the base64 encoding of the
 * object's underlying type (e.g. workinstance) and its uuid (i.e. primary
 * key).
 */
export type GlobalId = {
  type: string;
  id: string;
  suffix?: string[];
};

export function decodeGlobalId(id: unknown): GlobalId {
  return GlobalId.parse(decodeGlobalIdRaw(id));
}

export function tryDecodeGlobalId(
  id: unknown,
): { ok: true; value: GlobalId } | { ok: false; error: unknown } {
  try {
    return {
      ok: true as const,
      value: decodeGlobalId(id),
    };
  } catch (e) {
    return {
      ok: false as const,
      error: e,
    };
  }
}

export function decodeGlobalIdRaw(id: unknown): string {
  if (typeof id !== "string") {
    throw new Error(
      `invariant violated: global ids should be string but got ${typeof id}`,
    );
  }
  return Buffer.from(decodeURIComponent(id), "base64").toString();
}

export type GlobalIdInput = {
  type: string;
  id: string;
  suffix?: string | string[];
};

export function encodeGlobalId({ type, id, ...rest }: GlobalIdInput) {
  const suffix = rest.suffix
    ? typeof rest.suffix === "string"
      ? rest.suffix
      : rest.suffix.join(":")
    : "";
  return Buffer.from(
    `${type}:${id}${suffix.length ? `:${suffix}` : ""}`,
  ).toString("base64");
}

// biome-ignore lint/suspicious/noExplicitAny:
export function isGlobalId(value: any): value is GlobalId {
  if (typeof value?.type === "string" && typeof value?.id === "string") {
    return true;
  }

  return false;
}
