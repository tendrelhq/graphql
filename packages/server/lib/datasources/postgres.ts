import postgres, { type Fragment } from "postgres";
import z from "zod";

const { DB_MAX_CONNECTIONS, DB_STATEMENT_TIMEOUT_MS } = z
  .object({
    DB_MAX_CONNECTIONS: z.number({ coerce: true }).default(3),
    DB_STATEMENT_TIMEOUT_MS: z.number({ coerce: true }).default(10_000),
  })
  .parse(process.env);

export const sql = postgres({
  max: DB_MAX_CONNECTIONS,
  connection: {
    // We do some recursive queries and are, in general, not particularly
    // optimized so this is just a last resort to avoid bricking the backend.
    statement_timeout: DB_STATEMENT_TIMEOUT_MS,
  },
  types: {
    bigint: postgres.BigInt,
  },
});

export type Sql = typeof sql;
export type TxSql = Parameters<Parameters<typeof sql.begin>[1]>[0];

/**
 * Like Array.prototype.join but for sql.Fragments.
 */
export function join(xs: readonly Fragment[], d: Fragment) {
  return xs.reduce((acc, x, i) => sql`${acc} ${i ? sql`${d} ${x}` : x}`, sql``);
}

/**
 * Type predicate for whether to include a user input in a dynamic update
 * clause. Note that we only check for `undefined`, and not `null`, because we
 * say that `null` indicates we want to set the underlying database column to
 * NULL.
 */
export function shouldUpdate<T>(input?: T, existing?: T): input is T {
  return typeof input !== "undefined" && input !== existing;
}

export function unionAll(xs: readonly Fragment[]) {
  return join(xs, sql`UNION ALL`);
}
