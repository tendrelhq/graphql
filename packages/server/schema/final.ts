import { resolvers, typeDefs } from "@/schema";
import { makeExecutableSchema, mergeSchemas } from "@graphql-tools/schema";
import type { GraphQLScalarType } from "graphql";
import { type GqlUrl, register } from "./system/scalars/url";
import { getSchema } from "./v1.schema";

const legacySchema = makeExecutableSchema({ typeDefs, resolvers });
const gratsSchema = getSchema();

export const schema = mergeSchemas({ schemas: [legacySchema, gratsSchema] });

// Custom scalars. See https://grats.capt.dev/docs/docblock-tags/scalars/#serialization-and-parsing-of-custom-scalars.
register(schema.getType("URL") as GraphQLScalarType<GqlUrl, string>);
