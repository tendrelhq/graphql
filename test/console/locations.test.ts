import { afterAll, beforeAll, describe, expect, test } from "bun:test";
import { sql } from "@/datasources/postgres";
import { schema } from "@/schema/final";
import { decodeGlobalId, encodeGlobalId } from "@/schema/system";
import { execute, paginateQuery } from "@/test/prelude";
import {
  ListLocationsTestDocument,
  PaginateLocationsTestDocument,
} from "./locations.test.generated";

describe("[console] locations", () => {
  let ACCOUNT: string; // customer

  test("list locations", async () => {
    const result = await execute(schema, ListLocationsTestDocument, {
      account: ACCOUNT,
    });
    expect(result.errors).toBeFalsy();
    expect(result.data).toMatchSnapshot();
  });

  test("pagination", async () => {
    let i = 0;
    for await (const page of paginateQuery({
      async execute(cursor) {
        return await execute(schema, PaginateLocationsTestDocument, {
          account: ACCOUNT,
          first: 2,
          after: cursor,
        });
      },
      next(result) {
        if (result.data?.node.__typename !== "Organization") {
          throw "invariant violated";
        }
        return result.data.node?.locations.pageInfo;
      },
    })) {
      expect(page.errors).toBeFalsy();
      i++;
    }
    expect(i).toBe(3); // 6 total, 2 per page
  });

  test("search", async () => {
    const result = await execute(schema, ListLocationsTestDocument, {
      account: ACCOUNT,
      search: {
        active: true,
        isSite: true,
      },
    });
    expect(result.data?.node.__typename).toBe("Organization");
    if (result.data?.node.__typename === "Organization") {
      expect(result.data.node.locations.totalCount).toBe(1);
    }
  });

  beforeAll(async () => {
    const logs = await sql<{ op: string; id: string }[]>`
      select *
      from
          runtime.create_demo(
              customer_name := 'Frozen Tendy Factory',
              admins := (
                  select array_agg(workeruuid)
                  from public.worker
                  where workerfullname in (
                      'Jerry Garcia',
                      'Mike Heavner',
                      'Will Ruggiano',
                      'Will Twait'
                  )
              ),
              modified_by := 895
          )
      ;
    `;
    // we get customer uuid back in the first row
    const row1 = logs.at(0);
    // but we can check the tag to be sure
    if (row1?.op?.trim() !== "+customer") {
      throw "setup failed to find customer";
    }
    ACCOUNT = encodeGlobalId({
      type: "organization", // bleh
      id: row1.id,
    });
  });

  afterAll(async () => {
    const { id } = decodeGlobalId(ACCOUNT);
    process.stdout.write("Cleaning up... ");
    const [row] = await sql<[{ ok: string }]>`
      select runtime.destroy_demo(${id}) as ok;
    `;
    console.log(row.ok);
  });
});
