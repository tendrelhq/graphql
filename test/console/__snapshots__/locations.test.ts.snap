// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`[console] locations list locations 1`] = `
{
  "node": {
    "__typename": "Organization",
    "locations": {
      "__typename": "LocationConnection",
      "edges": [
        {
          "__typename": "LocationEdge",
          "node": {
            "__typename": "Location",
            "active": {
              "active": true,
            },
            "geofence": null,
            "name": {
              "value": "Frozen Tendy Factory",
            },
            "parent": null,
            "scanCode": null,
            "site": {
              "name": {
                "value": "Frozen Tendy Factory",
              },
            },
            "timeZone": "GMT",
          },
        },
        {
          "__typename": "LocationEdge",
          "node": {
            "__typename": "Location",
            "active": {
              "active": true,
            },
            "geofence": null,
            "name": {
              "value": "Mixing Line",
            },
            "parent": {
              "name": {
                "value": "Frozen Tendy Factory",
              },
            },
            "scanCode": null,
            "site": {
              "name": {
                "value": "Frozen Tendy Factory",
              },
            },
            "timeZone": "GMT",
          },
        },
        {
          "__typename": "LocationEdge",
          "node": {
            "__typename": "Location",
            "active": {
              "active": true,
            },
            "geofence": null,
            "name": {
              "value": "Fill Line",
            },
            "parent": {
              "name": {
                "value": "Frozen Tendy Factory",
              },
            },
            "scanCode": null,
            "site": {
              "name": {
                "value": "Frozen Tendy Factory",
              },
            },
            "timeZone": "GMT",
          },
        },
        {
          "__typename": "LocationEdge",
          "node": {
            "__typename": "Location",
            "active": {
              "active": true,
            },
            "geofence": null,
            "name": {
              "value": "Assembly Line",
            },
            "parent": {
              "name": {
                "value": "Frozen Tendy Factory",
              },
            },
            "scanCode": null,
            "site": {
              "name": {
                "value": "Frozen Tendy Factory",
              },
            },
            "timeZone": "GMT",
          },
        },
        {
          "__typename": "LocationEdge",
          "node": {
            "__typename": "Location",
            "active": {
              "active": true,
            },
            "geofence": null,
            "name": {
              "value": "Cartoning Line",
            },
            "parent": {
              "name": {
                "value": "Frozen Tendy Factory",
              },
            },
            "scanCode": null,
            "site": {
              "name": {
                "value": "Frozen Tendy Factory",
              },
            },
            "timeZone": "GMT",
          },
        },
        {
          "__typename": "LocationEdge",
          "node": {
            "__typename": "Location",
            "active": {
              "active": true,
            },
            "geofence": null,
            "name": {
              "value": "Packaging Line",
            },
            "parent": {
              "name": {
                "value": "Frozen Tendy Factory",
              },
            },
            "scanCode": null,
            "site": {
              "name": {
                "value": "Frozen Tendy Factory",
              },
            },
            "timeZone": "GMT",
          },
        },
      ],
      "pageInfo": {
        "__typename": "PageInfo",
        "hasNextPage": false,
        "hasPreviousPage": true,
      },
      "totalCount": 6,
    },
  },
}
`;
